nx/source/services/ncm.c

```diff
@@ -1,1029 +1,484 @@
-    raw->index = index;
-    memcpy(&raw->meta_record, record, sizeof(NcmMetaRecord));
-    
-    Result rc = serviceIpcDispatch(&db->s);
-    if (R_SUCCEEDED(rc)) {
-        IpcParsedCommand r;
-        ipcParse(&r);
+void ncmContentMetaDatabaseClose(NcmContentMetaDatabase* db) {
+    serviceClose(&db->s);
+}
 
-        struct {
-            u64 magic;
-            u64 result;
-            u32 entries_read;
-        } *resp = r.Raw;
+Result ncmContentMetaDatabaseSet(NcmContentMetaDatabase* db, const NcmContentMetaKey* key, const void* data, u64 data_size) {
+    return serviceDispatchIn(&db->s, 0, *key,
+        .buffer_attrs = { SfBufferAttr_HipcMapAlias | SfBufferAttr_In },
+        .buffers = { { data, data_size } },
+    );
+}
 
-        rc = resp->result;
+Result ncmContentMetaDatabaseGet(NcmContentMetaDatabase* db, const NcmContentMetaKey* key, u64* out_size, void* out_data, u64 out_data_size) {
+    return serviceDispatchInOut(&db->s, 1, *key, *out_size, 
+        .buffer_attrs = { SfBufferAttr_HipcMapAlias | SfBufferAttr_Out },
+        .buffers = { { out_data, out_data_size } },
+    );
+}
 
-        if (R_SUCCEEDED(rc)) {
-            if (numEntriesRead) *numEntriesRead = resp->entries_read;
-        }
-    }
-    
-    return rc;
+Result ncmContentMetaDatabaseRemove(NcmContentMetaDatabase* db, const NcmContentMetaKey *key) {
+    return serviceDispatchIn(&db->s, 2, *key);
 }
 
-Result ncmContentMetaDatabaseList(NcmContentMetaDatabase* db, u32 titleType, u64 titleIdExact, u64 titleIdLow, u64 titleIdHigh, NcmMetaRecord* metaRecordsOut, size_t metaRecordsBufSize, u32* numEntriesWritten, u32* numEntriesTotal) {
-    IpcCommand c;
-    ipcInitialize(&c);    
-    ipcAddRecvBuffer(&c, metaRecordsOut, metaRecordsBufSize, BufferType_Normal);
-    
-    struct {
-        u64 magic;
-        u64 cmd_id;
-        u32 titleType;
-        u64 titleIdExact;
-        u64 titleIdLow;
-        u64 titleIdHigh;
-    } *raw;
-    
-    raw = ipcPrepareHeader(&c, sizeof(*raw));
-    
-    raw->magic = SFCI_MAGIC;
-    raw->cmd_id = 5;
-    raw->titleType = titleType;
-    raw->titleIdExact = titleIdExact;
-    raw->titleIdLow = titleIdLow;
-    raw->titleIdHigh = titleIdHigh;
-    
-    Result rc = serviceIpcDispatch(&db->s);
-    if (R_SUCCEEDED(rc)) {
-        IpcParsedCommand r;
-        ipcParse(&r);
-
-        struct {
-            u64 magic;
-            u64 result;
-            u32 numEntriesTotal;
-            u32 numEntriesWritten;
-        } *resp = r.Raw;
-
-        rc = resp->result;
-
-        if (R_SUCCEEDED(rc)) {
-            if (numEntriesTotal)
-                *numEntriesTotal = resp->numEntriesTotal;
-            if (numEntriesWritten)
-                *numEntriesWritten = resp->numEntriesWritten;
-        }
-    }
-    
-    return rc;
+Result ncmContentMetaDatabaseGetContentIdByType(NcmContentMetaDatabase* db, NcmNcaId* out_content_id, const NcmContentMetaKey* key, NcmContentType type) {
+    const struct {
+        u32 type;
```

Comment:
Is this correct?(wiki has u8 for this)

Reply:
I left the types for this struct the same as they are currently, however, technically this is a u8 type + 0x7 in padding. I'll change it for clarity.